test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,391 @@
+
+5) Create agnHost 2 pods (deployment?).
+6) Get IP of one of the pods of agnHost.
+7) create and configure the custom CoreDNS configMap with the IP received above.
+8) Create a custom DNS pod and configure it to use above configMap with agnHost pod's IP.
+9) Create and apply antrea FQDN policy.
+10) Deploy antrea-toolbox.
+
+11) curl the FQDN from within toolbox.
+12) imitate caching the IP belonging to above FQDN resolution by keeping it in a variable.
+13) change the existing IP in dns configmap with the other agnHost pod's IP.
+14) wait for new IP to get updated in configMap and let the changes be reflected in dns pod.
+15) curl the FQDN again with old IP , simulating usage of cache -- and it must fail with no connectivity.
+*/
+
+// agnHostPodIpv4Addresses type stores the IPv4 addresses of the 2 agnHost pods created for the test.
+type agnHostPodIpv4Addresses struct {
+	pod1Ipv4 string
+	pod2Ipv4 string
+}
+
+func TestFQDNPolicyWithCachedDNS(t *testing.T) {
+	const testFQDN = "fqdn-test-pod.lfx.test"
+
+	skipIfAntreaPolicyDisabled(t)
+	skipIfNotIPv4Cluster(t)
+	skipIfIPv6Cluster(t)
+
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	dnsServiceIP := setCustomDnsServerAddressInAntrea(t, data, false)
+	defer setCustomDnsServerAddressInAntrea(t, data, true)
+
+	agnHostPodIp := createAgnHostPods(t, data)
+
+	customDnsConfigMap, err := createAndDeployCustomDnsConfigMap(t, data, agnHostPodIp.pod1Ipv4, testFQDN)
+	require.NoError(t, err, "failed to create custom dns ConfigMap: %v", err)
+
+	createDnsPod(t, data)
+	createFqdnPolicyInNamespace(t, data, testFQDN)
+
+	t.Logf("Creating Toolbox pod...")
+	createToolBoxPod(t, data, dnsServiceIP)
+
+	fqdnIp, err := curlAndVerifyFQDN(t, data, testFQDN, dnsServiceIP)
+	require.NoError(t, err, "failed to resolve FQDN to an IP from toolbox pod : %v", err)
+	require.Equalf(t, agnHostPodIp.pod1Ipv4, fqdnIp, "The IP set against the FQDN in the DNS server should be the same, but got %s instead of %s", fqdnIp, agnHostPodIp.pod1Ipv4)
+
+	t.Logf("Successfully received the expected IP %s using the dig command against the FQDN", fqdnIp)
+
+	t.Logf("New IP to update to DNS | ipv4 : %s", agnHostPodIp.pod2Ipv4)
+	err = updateCustomDnsConfigMap(t, data, agnHostPodIp.pod2Ipv4, testFQDN, customDnsConfigMap)
+	require.NoError(t, err, "failed to update configmap with new IP : %v", err)
+	t.Logf("successfully updated dns configMap with new IPs | ipv4 : %s", agnHostPodIp.pod2Ipv4)
+
+	require.NoError(t, data.setPodAnnotation(data.testNamespace, "custom-dns-server", randomPatchAnnotationKey,
+		randSeq(annotationValueLen)), "failed to update custom dns pod annotation.")
+
+	cachedIpTest(t, fqdnIp, data)
+}
+
+// createAgnHostPods creates two agnHost pods and returns their IPv4 addresses filled in agnHostPodIpv4Addresses type.
+func createAgnHostPods(t *testing.T, data *TestData) *agnHostPodIpv4Addresses {
```

Comment:
I think this function and the `agnHostPodIpv4Addresses` struct are hurting readability and are not necessary: you already have a `createHttpAgnhostPod` helper function, you can move creation to the main test function.

Reply:
Even `createHttpAgnhostPod` is probably not necessary, but we can keep it for now for the sake of moving forward.